### PR TITLE
Add ClearableManagerInterface

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Added `ClearableManagerInterface` to be able to clear the Manager in an easy way.
```

### CODE REVIEW

**Comment by jordisala1991** [comment]
I will need a release after this one to keep going on MediaBundle https://github.com/sonata-project/SonataMediaBundle/pull/1975

File: src/Model/ClearableManagerInterface.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\Doctrine\Model;
+
+/**
+ * @author Jordi Sala <jordism91@gmail.com>
+ *
+ * @phpstan-template T of object
```
Comment (VincentLanglet) [inline src/Model/ClearableManagerInterface.php:19]: I don't think you need this since it's not used.

Comment (jordisala1991) [reply src/Model/ClearableManagerInterface.php]: Copy paste mistake, fixed.

**Review by VincentLanglet — APPROVED** [review]

**Review by core23 — COMMENTED** [review]
Do we really need this as a common extension point? Looks like custom business logic.

Never used this methods.

**Comment by VincentLanglet** [comment]
> Do we really need this as a common extension point? Looks like custom business logic.
> 
> Never used this methods.

Yes, we need something like that.

It's used for instance in the SonataMediaBundle
https://github.com/sonata-project/SonataMediaBundle/blob/1ba904a8d6c0330d63b05de23c9fa8386fded41f/src/Command/SyncThumbsCommand.php#L143

`getObjectManager()` does not exist on the interface.
It seems better to expose a clear method instead in an interface, allowing to do
```
if ($this->mediaManager instanceof ClearableManagerInterface) {
     $this->mediaManager->clear();
}
```

**Comment by VincentLanglet** [comment]
@jordisala1991 We should also add a NEXT_MAJOR comment to make `getObjectManager` protected.

**Comment by jordisala1991** [comment]
> @jordisala1991 We should also add a NEXT_MAJOR comment to make `getObjectManager` protected. (And all the public method of the AbstractManager which are not in the interface).

Just added a new issue to tackle that

**Comment by jordisala1991** [comment]
Just found another use case for this: https://github.com/sonata-project/SonataMediaBundle/blob/3.x/src/Command/AddMassMediaCommand.php#L122-L125

File: src/Adapter/ORM/DoctrineORMAdapter.php [hunk 2]
```diff
@@ -32,27 +32,27 @@ public function __construct(ManagerRegistry $registry)
...
-    public function getNormalizedIdentifier($entity)
+    public function getNormalizedIdentifier($model)
     {
-        if (null === $entity) {
+        if (null === $model) {
             return null;
         }
 
-        if (!\is_object($entity)) {
+        if (!\is_object($model)) {
```
Comment (phansys) [inline src/Adapter/ORM/DoctrineORMAdapter.php:41]: We should move this check to a type declaration in the next major.

File: src/Model/ClearableManagerInterface.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Sonata\Doctrine\Model;
+
+/**
+ * @author Jordi Sala <jordism91@gmail.com>
+ */
+interface ClearableManagerInterface
+{
+    /**
+     * Clears the object manager.
```
Comment (phansys) [inline src/Model/ClearableManagerInterface.php:22]:
```suggestion
     * Clears the object manager for the given model.
```

**Review by phansys — APPROVED** [review]

File: src/Adapter/ORM/DoctrineORMAdapter.php (hunk 2 above)
Comment (jordisala1991) [reply src/Adapter/ORM/DoctrineORMAdapter.php:41]: Agree, this was only changed to make psalm/phpstan happy

File: src/Model/ClearableManagerInterface.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Sonata\Doctrine\Model;
+
+/**
+ * @author Jordi Sala <jordism91@gmail.com>
+ */
+interface ClearableManagerInterface
+{
+    /**
+     * Clears the object manager for the given model.
```
Comment (phansys) [inline src/Model/ClearableManagerInterface.php:22]: Checking the implementation, I wonder what happens if the argument value is `null`? The whole manager is cleared? If that is the case, I think we should make the behavior clear in this description.

**Review by phansys — APPROVED** [review]

File: src/Model/ClearableManagerInterface.php (hunk 4 above)
Comment (jordisala1991) [reply src/Model/ClearableManagerInterface.php:22]: This null behavior is copied from the original doctrine method: https://github.com/doctrine/persistence/blob/2.2.x/lib/Doctrine/Persistence/ObjectManager.php#L69-L77

**Review by VincentLanglet — APPROVED** [review]

File: src/Model/ClearableManagerInterface.php (hunk 4 above)
Comment (phansys) [reply src/Model/ClearableManagerInterface.php:22]: IMO, we could use a similar description for the method and the argument then.

**Pull request merged**